cylc/flow/data_store_mgr.py

```diff
@@ -539,19 +542,178 @@ def generate_definition_elements(self):
+        # associated with the current Active node.
+        if edge_distance > self.n_edge_distance:
+            if descendant or edge_distance == 0:
+                self.n_window_boundary_nodes[
+                    active_id].setdefault(edge_distance, set()).add(s_id)
+            return
+        graph_children = generate_graph_children(
+            self.schd.config.get_taskdef(name), point)
+        if not any(graph_children.values()):
+            if descendant or edge_distance == 0:
+                self.n_window_boundary_nodes[
+                    active_id].setdefault(edge_distance, set()).add(s_id)
+
+        self.n_window_nodes[active_id].add(s_id)
+        # Generate task node
+        self.generate_ghost_task(s_id, name, point, flow_label)
+
+        edge_distance += 1
+
+        # TODO: xtrigger is suite_state edges too
+        # Reference set for workflow relations
+        for items in graph_children.values():
+            if edge_distance == 1:
+                descendant = True
+            self._expand_graph_window(
+                s_id, s_node, items, active_id, flow_label, edge_distance,
+                descendant, False)
+
+        for items in generate_graph_parents(
+                self.schd.config.get_taskdef(name), point).values():
+            self._expand_graph_window(
+                s_id, s_node, items, active_id, flow_label, edge_distance,
+                False, True)
+
+        if edge_distance == 1:
+            levels = self.n_window_boundary_nodes[active_id].keys()
+            # Could be self-reference node foo:failed => foo
+            if not levels:
+                self.n_window_boundary_nodes[active_id][0] = {active_id}
+                levels = (0,)
+            # Only trigger pruning for futhest set of boundary nodes
+            for tp_id in self.n_window_boundary_nodes[active_id][max(levels)]:
+                self.prune_trigger_nodes.setdefault(
+                    tp_id, set()).add(active_id)
+            del self.n_window_boundary_nodes[active_id]
+            if self.n_window_edges[active_id]:
+                getattr(self.updated[WORKFLOW], EDGES).edges.extend(
+                    self.n_window_edges[active_id])
+
+    def _expand_graph_window(
+            self, s_id, s_node, items, active_id, flow_label, edge_distance,
+            descendant=False, is_parent=False):
+        """Construct nodes/edges for children/parents of source node."""
+        for t_name, t_point, is_abs in items:
+            t_node = TaskID.get(t_name, t_point)
+            t_id = (
+                f'{self.workflow_id}{ID_DELIM}{t_point}{ID_DELIM}{t_name}')
+            # Initiate edge element.
+            if is_parent:
+                e_id = (
+                    f'{self.workflow_id}{ID_DELIM}{t_node}{ID_DELIM}{s_node}')
+            else:
+                e_id = (
+                    f'{self.workflow_id}{ID_DELIM}{s_node}{ID_DELIM}{t_node}')
```

Comment:
I'm confused by these lines:

* The `{s,t}_node`s are in `<task>.<cycle>` format but being templated into a universal ID?
* The first `{s,t}_node` is going to be interpreted as a cycle point?

Reply:
Taskdef ID is `{self.workflow_id}{ID_DELIM}{name}` .. So there's no promise of the the third item being a cycle point.. Whether there should be is a different question.. Maybe?   

> <task>.<cycle> format but being templated into a universal ID

Well the source and target needs to be unique in the edge ID ... Can't be just the name as we have intercycle dependencies.